### DISCUSSION TITLE
How to obtain/keep old value in onupdate default?

### DISCUSSION THREAD

**Discussion by jerem1e** [sqlalchemy/sqlalchemy#10572] [Usage Questions]
I have the following Column:

```python
    last_update_step = Column(
        String, nullable=False, onupdate=default_update_step, default=default_update_step
    )
```

And this is the `default_update_step` function:
```python
def default_update_step(context: SQLiteExecutionContext_aiosqlite) -> str:
    params = context.get_current_parameters()
    for key in params.keys():
        if key in [
            consts.FIRST_STEP,
            consts.OTHER_STEP,
            consts.NEW_STEP,
        ]:
            return key
```

In the case, I don't find any of the results in parameters I want to return / obtain the old value of `last_update_step`.
How can I do that?
Any help would be much appreciated.

**Comment by zzzeek** [comment]
the contents of `context.get_current_parameters()` depends upon the specific INSERT or UPDATE statement being emitted.

If you need general information about an ORM object when firing off this default I would consider using ORM-level hooks for this, such as the [before_insert](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.MapperEvents.before_insert) and [before_update](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.MapperEvents.before_update) hooks.

**Reply by jerem1e** [reply]
`before_insert`and `before_update` don't work for me as I use query updates...

**Reply by zzzeek** [reply]
if you emit an UPDATE statement that is not updating any of those columns `consts.FIRST_STEP`, `consts.OTHER_STEP`, `consts.NEW_STEP`, then they wont be in the params.

I dont really understand what these `consts.FIRST_STEP` are, the way they are being used suggests they match the names of columns, which would be an odd programming pattern.
